pkg/controller/templateupdaterequest/templateupdaterequest_controller.go

```diff
@@ -0,0 +1,320 @@
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, _ *configuration.Config) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileTemplateUpdateRequest{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("templateupdaterequest-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource TemplateUpdateRequest
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.TemplateUpdateRequest{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileTemplateUpdateRequest implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileTemplateUpdateRequest{}
+
+// ReconcileTemplateUpdateRequest reconciles a TemplateUpdateRequest object
+type ReconcileTemplateUpdateRequest struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a TemplateUpdateRequest object and makes changes based on the state read
+// and what is in the TemplateUpdateRequest.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileTemplateUpdateRequest) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	logger.Info("Reconciling TemplateUpdateRequest")
+
+	// Fetch the TemplateUpdateRequest tur
+	tur := &toolchainv1alpha1.TemplateUpdateRequest{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, tur)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// if is complete, then check when status was changed and delete it if the requested duration has passed
+	completeCond, found := condition.FindConditionByType(tur.Status.Conditions, toolchainv1alpha1.TemplateUpdateRequestComplete)
+	if found && completeCond.Status == corev1.ConditionTrue {
+		deleted, requeueAfter, err := r.checkTransitionTimeAndDelete(logger, tur, completeCond)
```

Comment:
does it really make such a big difference? it works fine in the e2e tests this way.

Reply:
The difference is who is responsible for what - as I see it, the `NSTemplateTier` controller is responsible for `TemplateUpdateRequest` CRs:
* it creates them
* it watches them and tracks them
* and it should also delete them when the controller decides that it's the right time

Reply:
Yes, the NSTemplateTierController watches the `NSTemplateTier` as its primary resources and that's the kind of resource that it loads during each reconcile loop. 
Are you suggesting that each time a reconcile loop is triggered, it should also list all `TemplateUpdateRequest` that are in `ready=true/reason=updated` condition?

Reply:
> Yes, the NSTemplateTierController watches the NSTemplateTier as its primary resources and that's the kind of resource that it loads during each reconcile loop.

Yes, and the NSTemplateTierController also watches the `TemplateUpdateRequest` CRD https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/nstemplatetier/nstemplatetier_controller.go#L63-L67 so the reconcile is triggered for every update of any of the `TemplateUpdateRequest` CRs

> Are you suggesting that each time a reconcile loop is triggered, it should also list all TemplateUpdateRequest that have been in ready=true/reason=updated condition

Yes, and not only `complete=true/reason=updated` but also `complete=false/reason=updateFailed`. This is the only way to know how many updates have passed and which updates have failed. 
* The list of the failed updates is necessary for later troubleshooting
* The number of the successful updates is handy to know what is the progress of the update and to guess how many updates is still missing. 

> for what, a few seconds (or immediately)?

* for the successful ones we can delete them immediately
* for the failed ones, I would wait 3 seconds to give some time to the controller to fix the problem without manual cooperation (eg. for the cases when it wasn't able to update the MUR because it has been already updated, so the controller didn't have the latest data)

Reply:
> for the failed ones, I would wait 3 seconds to give some time to the controller to fix the problem without manual cooperation (eg. for the cases when it wasn't able to update the MUR because it has been already updated, so the controller didn't have the latest data)

that seems like a weak way. We should set the TemplateUpdateRequest in such a condition that the controller will not remove it at all. And then, after a few attempts, in a condition that gives "permission" to the controller to delete it. Bu that does not say how we want to trace the failures.

Reply:
also, what worries me is that we're going to list (fetch) all TemplateUpdateRequest resources each time one of them is updated, since there's no way to filter by status condition, I believe :(

Reply:
> that seems like a weak way. We should set the TemplateUpdateRequest in such a condition that the controller will not remove it at all. And then, after a few attempts, in a condition that gives "permission" to the controller to delete it.

yes, this approach is even better - my point was to not delete it immediately and unless we don't have https://issues.redhat.com/browse/CRT-708 implemented, that approach seemed to be the easiest one. Anyway, if you are going to work on it soon then there is no need for such a temporal solution, indeed :+1:

Reply:
> also, what worries me is that we're going to list (fetch) all TemplateUpdateRequest resources each time one of them is updated, since there's no way to filter by status condition, I believe :(

no, but I don't expect that we would have many of TemplateUpdateRequest CRs - right now the maximum is 10, which is fine. 
and btw, the controller is listing all TemplateUpdateRequest resources each time anyway https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/nstemplatetier/nstemplatetier_controller.go#L185-L188

Reply:
In general I like the approach when a Primary Resource controller manages the entire cycle of the Secondary resources. So, Tier controller creates, deletes and updates UpdateRequests.
But I wouldn't take it as a must rule. Sometimes it's more practical to delegate "self-destruction/cleanup" to the secondary resource controller as we did for Notifications (there can be multiple clients/primary controllers) and ChangeTierRequests (there is no clients/primary controllers at all)

So, in this particular case, if Tier controller needs to track the lifecycle of the UpdateRequests (to update Tier status for example) then IMO it makes sense to delete UpdateRequests by the Tier controller. Otherwise such tracking would be quite problematic. This is not the case for Notification or ChangeTierRequest.

Reply:
@MatousJobanek @alexeykazakov ok, I understand the pattern here and I will apply the changes in this PR accordingly. The NSTemplateTier controller will list all `TemplateUpdateRequest` resources on each reconcile loop and delete those that reached a "terminating" state (ie, `Ready=true/Reason=Updated` or `Ready=false/Reason=UpdateFailed`.

Reply:
done in fddc0ab